### PR TITLE
feat(v2/storage-3): allow form admins to download storage mode responses

### LINKED ISSUE #3129: [React/StorageMode] Download storage mode form responses
Note that even if #2791 is not completed we can mock something light to allow for the download of responses

### PR DESCRIPTION
**Pull request by karrui**
Third PR of chain implementing download feature. Builds on #3950.

## Problem

This PR allows form admins and collaborator to download storage mode responses

Closes #3129

## Solution

**Breaking Changes** 

- [x] No - this PR is backwards compatible  

**Features**:

- feat: add useSecretKey hook with broadcast channel capabilities
- feat: add download button to download responses when clicked

**Improvements**:

- ref: rename ResponsesProvider to StorageResponsesProvider

## Before & After Screenshots
ResponseTab should show new (non-working) download button

## Deploy Notes

**New dependencies**:

- `broadcast-channel` : used for broadcasting secret key across tabs

### CODE REVIEW

File: frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadButton.tsx [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  DownloadEncryptedParams,
+} from '../useDecryptionWorkers'
+
+export const DownloadButton = (): JSX.Element => {
+  const { downloadEncryptedResponses } = useDecryptionWorkers()
+  const { downloadParams } = useStorageResponsesContext()
+
+  const handleExportCsvMutation = useMutation(
+    (params: DownloadEncryptedParams) => downloadEncryptedResponses(params),
+    // TODO: add error and success handling
```
Comment (mantariksh) [inline frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadButton.tsx:22]: is this happening in a subsequent PR?

File: frontend/src/features/admin-form/responses/ResponsesPage/storage/useDecryptionWorkers.ts [hunk 2]
```diff
@@ -27,18 +27,21 @@ export type DownloadEncryptedParams = EncryptedResponsesStreamParams & {
...
   const { data: adminForm } = useAdminForm()
 
   const { refetch } = useFormResponsesCount()
 
   useEffect(() => {
-    const abortController = abortControllerRef.current
+    return () => killWorkers(workers)
+  }, [workers])
 
+  useEffect(() => {
```
Comment (mantariksh) [inline frontend/src/features/admin-form/responses/ResponsesPage/storage/useDecryptionWorkers.ts:39]: why use two separate `useEffect`s?

File: frontend/src/features/admin-form/responses/ResponsesPage/storage/useSecretKey.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+// BroadcastChannel will only broadcast the message to scripts from the same origin
+// (i.e. https://form.gov.sg in practice) so all data should be controlled by scripts
+// originating from FormSG. This does not store any data in browser-based storage
+// (e.g. cookies or localStorage) so secrets would not be retained past the user closing
+// all Form tabs containing the form.
+const secretKeyChannel = new BroadcastChannel<SecretKeyBroadcastMessage>(
+  SECRETKEY_BROADCAST_KEY,
+)
+
+export const useSecretKey = (formId: string) => {
```
Comment (mantariksh) [inline frontend/src/features/admin-form/responses/ResponsesPage/storage/useSecretKey.ts:28]: this is really just for broadcast right? maybe `useSecretKeyBroadcast` to make the purpose clear

**Review by mantariksh — APPROVED** [review]
question on the error handling

File: frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadButton.tsx (hunk 1 above)
Comment (karrui) [reply frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadButton.tsx:22]: yes

File: frontend/src/features/admin-form/responses/ResponsesPage/storage/useDecryptionWorkers.ts (hunk 2 above)
Comment (karrui) [reply frontend/src/features/admin-form/responses/ResponsesPage/storage/useDecryptionWorkers.ts:39]: one is dependent on workers, the other isn't haha.

File: frontend/src/features/admin-form/responses/ResponsesPage/storage/useSecretKey.ts (hunk 3 above)
Comment (karrui) [reply frontend/src/features/admin-form/responses/ResponsesPage/storage/useSecretKey.ts:28]: also to generate and keep state of secret key, so i think i'll keep the name

**Pull request merged**